# Cherry Blossoms!

If we travel back in time, [cherry blossoms](https://en.wikipedia.org/wiki/Cherry_blossom) were once in full bloom! We don't live in Japan or DC, but we do have the [Brooklyn Botanic Garden's annual festival](https://www.bbg.org/visit/event/sakura_matsuri_2020) (whenever COVID isn't ruining things, at least).

We'll have to make up for it with data-driven cherry blossoms instead. Once upon a time [Data is Plural](https://tinyletter.com/data-is-plural) linked to [a dataset](http://atmenv.envi.osakafu-u.ac.jp/aono/kyophenotemp4/) about when the cherry trees blossom each year. It's completely out of date, but it's quirky in a real nice way so we're sticking with it.

## 0. Do all of your importing/setup stuff

In [5]:
import pandas as pd
#! pip install xlrd

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.5/96.5 kB 1.9 MB/s eta 0:00:00a 0:00:01


## 1. Read in the file using pandas, and look at the first five rows

In [20]:
df = pd.read_excel("KyotoFullFlower7.xls", skiprows=25, na_values=['-'])
df.head(5)

AD  Full-flowering date (DOY)  Full-flowering date  Source code  \
0  801                        NaN                  NaN          NaN   
1  802                        NaN                  NaN          NaN   
2  803                        NaN                  NaN          NaN   
3  804                        NaN                  NaN          NaN   
4  805                        NaN                  NaN          NaN   

   Data type code Reference Name  
0             NaN            NaN  
1             NaN            NaN  
2             NaN            NaN  
3             NaN            NaN  
4             NaN            NaN

## 2. Read in the file using pandas CORRECTLY, and look at the first five rows

Hrm, how do your column names look? Read the file in again but this time add a parameter to make sure your columns look right.


* **TIP:** Look at the documentation for `read_excel` and `read_csv` to see if there's anything that allows you to skip rows*
* **TIP:** The first year should be 801 AD, and it should not have any dates or anything.

In [21]:
df.columns

Index(['AD', 'Full-flowering date (DOY)', 'Full-flowering date', 'Source code',
       'Data type code', 'Reference Name'],
      dtype='object')

## 3. Look at a random selection of 10 rows of data

In [22]:
df.sample(10)

AD  Full-flowering date (DOY)  Full-flowering date  Source code  \
471   1272                        NaN                  NaN          NaN   
6      807                        NaN                  NaN          NaN   
1022  1823                      106.0                416.0          3.0   
1165  1966                       97.0                407.0          3.0   
335   1136                        NaN                  NaN          NaN   
683   1484                       97.0                406.0          1.0   
945   1746                      110.0                420.0          4.0   
288   1089                        NaN                  NaN          NaN   
35     836                        NaN                  NaN          NaN   
454   1255                        NaN                  NaN          NaN   

      Data type code          Reference Name  
471              NaN                     NaN  
6                NaN                     NaN  
1022             2.0         RAI-SANYO-NIPPU  
1165             0.0  NEWS-PAPER(ARASHIYAMA)  
335              NaN                     NaN  
683              2.0    OYUDONONO-UENO-NIKKI  
945              2.0             MICHIE-KOKI  
288              NaN                     NaN  
35               NaN                     NaN  
454              NaN                     NaN

## 4. Add some more NaN values

It looks like you should probably have some NaN/missing values earlier on in the dataset under "Reference name." Read in the file *one more time*, this time making sure all of those missing reference names actually show up as `NaN` instead of `-`.

I added `na_values=['-']` in the command where I read the data.

Also before getting started, I clean up the column names:

In [24]:
df.columns = df.columns.str.lower().str.replace(" ","_")
df

ad  full-flowering_date_(doy)  full-flowering_date  source_code  \
0      801                        NaN                  NaN          NaN   
1      802                        NaN                  NaN          NaN   
2      803                        NaN                  NaN          NaN   
3      804                        NaN                  NaN          NaN   
4      805                        NaN                  NaN          NaN   
...    ...                        ...                  ...          ...   
1210  2011                       99.0                409.0          8.0   
1211  2012                      101.0                410.0          8.0   
1212  2013                       93.0                403.0          8.0   
1213  2014                       94.0                404.0          8.0   
1214  2015                       93.0                403.0          8.0   

      data_type_code          reference_name  
0                NaN                     NaN  
1                NaN                     NaN  
2                NaN                     NaN  
3                NaN                     NaN  
4                NaN                     NaN  
...              ...                     ...  
1210             0.0  NEWS-PAPER(ARASHIYAMA)  
1211             0.0  NEWS-PAPER(ARASHIYAMA)  
1212             0.0  NEWS-PAPER(ARASHIYAMA)  
1213             0.0  NEWS-PAPER(ARASHIYAMA)  
1214             0.0  NEWS-PAPER(ARASHIYAMA)  

[1215 rows x 6 columns]

## 4. What source is the most common as a reference?

In [26]:
df.source_code.value_counts()

# The most common source as a reference is 3, which according to the spreadsheet is:

# " 3; Added by Aono and Omoto (1994), J. Agric. Meteorol., 49, 263-272."

3.0    342
4.0    250
1.0    105
5.0     59
7.0     40
2.0     17
6.0      9
8.0      5
Name: source_code, dtype: int64

## 6. Filter the list to only include columns where the `Full-flowering date (DOY)` is not missing

If you'd like to do it in two steps (which might be easier to think through), first figure out how to test whether a column is empty/missing/null/NaN, get the list of `True`/`False` values, and then later feed it to your `df`.

## 7. Make a histogram of the full-flowering date

## 8. Make another histogram of the full-flowering date, but with 39 bins instead of 10

## 9a. What's the average number of days it takes for the flowers to blossom? 

## 9b. And how many records do we have?

You can actually answer this one and the last one with a single line of code.

## 10. How many days into the year did cherry flowers normally blossom before 1900?



## 11. How about after 1900?

## 12. How many times was our data from a title in Japanese poetry?

You'll need to read the documentation inside of the Excel file.

## 13. Show only the years where our data was from a title in Japanese poetry

## 14. Graph the full-flowering date (DOY) over time

## 15. Smooth out the graph

It's so jagged! You can use `df.rolling` to calculate a rolling average.

The following code calculates a **10-year mean**, using the `AD` column as the anchor. If there aren't 20 samples to work with in a row, it'll accept down to 5. Neat, right?

```python
df.rolling(10, on='AD', min_periods=5)['Full-flowering date (DOY)'].mean().tail()
```

In this case we're only looking at the final 5. You'll probably want to look at more!

Use the code above to create a new column called `rolling_date` in our dataset. It should be the 20-year rolling average of the flowering date. Then plot it, with the year on the x axis and the day of the year on the y axis.

Try adding `ylim=(80, 120)` to your `.plot` command to make things look a little less dire.

### 16. Add a month column

Right now the "Full-flowering date" column is pretty rough. It uses numbers like '402' to mean "April 2nd" and "416" to mean "April 16th." Let's make a column to explain what month it happened in.

* Every row that happened in April should have 'April' in the `month` column.
* Every row that happened in March should have 'March' as the `month` column.
* Every row that happened in May should have 'May' as the `month` column.

There are **at least two ways to do this.**

#### WAY ONE: The bad-yet-simple way

If you don't want to use `pd.to_datetime`, you can use this as an sample for updating March. It finds everything with a date less than 400 and assigns `March` to the `month` column:

```python
df.loc[df['Full-flowering date'] < 400, 'month'] = 'March'
```

#### WAY TWO: The good-yet-complicated way

* When you use `pd.to_datetime` pandas tries to figure out the date format automatically. If that doesn't work, you can also pass a `format=` argument that explains what the format is of the datetime. You use [the codes here](https://strftime.org/) to mark out where the days, months, etc are. For example, `2020-04-09` would be converted using `pd.to_datetime(df.colname, "format='%Y-%m-%d")`.
* `errors='coerce'` will return `NaN` for missing values.
* And remember how we used `df.date_column.dt.month` to get the number of the month? For the name, you use `dt.strftime` (string-formatted-time), and pass it [the same codes](https://strftime.org/) to tell it what to do. For example, `df.date_column.dt.strftime("%Y-%m-%d")` would give you `"2020-04-09"`.

### 17. Using your new column, how many blossomings happened in each month?

### 18. Graph how many blossomings happened in each month.

### 19. Adding a day-of-month column

Now we're going to add a new column called `day_of_month.` Depending on what you did above, it might be a little tougher since the `Full-flowering date` column is a *float* instead of an integer.

*Tip: If your method involves `.astype(int)` it isn't going to work since it's missing data, you can add `.dropna().astype(int)` instead.*

### 20. Adding a date column

If you don't have one yet, take the `'month'` and `'day_of_month'` columns and combine them in order to create a new string-like column called `'date'`. You could alternatively use `.dt.strftime` as mentioned above, or *just totally skip this one if you have an actual datetime column*.

# YOU ARE DONE.

And **incredible.**